src/DefaultBuilder/src/WebApplication.cs

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Hosting.Server;
+using Microsoft.AspNetCore.Hosting.Server.Features;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Builder
+{
+    /// <summary>
+    /// The web application used to configure the http pipeline, and routes.
+    /// </summary>
+    public sealed class WebApplication : IHost, IApplicationBuilder, IEndpointRouteBuilder, IAsyncDisposable
+    {
+        internal const string EndpointRouteBuilder = "__EndpointRouteBuilder";
+
+        private readonly IHost _host;
+        private readonly List<EndpointDataSource> _dataSources = new();
+
+        internal WebApplication(IHost host)
+        {
+            _host = host;
+            ApplicationBuilder = new ApplicationBuilder(host.Services);
+            Logger = host.Services.GetRequiredService<ILoggerFactory>().CreateLogger(Environment.ApplicationName);
+        }
+
+        /// <summary>
+        /// The application's configured services.
+        /// </summary>
+        public IServiceProvider Services => _host.Services;
+
+        /// <summary>
+        /// The application's configured <see cref="IConfiguration"/>.
+        /// </summary>
+        public IConfiguration Configuration => _host.Services.GetRequiredService<IConfiguration>();
+
+        /// <summary>
+        /// The application's configured <see cref="IWebHostEnvironment"/>.
+        /// </summary>
+        public IWebHostEnvironment Environment => _host.Services.GetRequiredService<IWebHostEnvironment>();
+
+        /// <summary>
+        /// Allows consumers to be notified of application lifetime events.
+        /// </summary>
+        public IHostApplicationLifetime Lifetime => _host.Services.GetRequiredService<IHostApplicationLifetime>();
+
+        /// <summary>
+        /// The default logger for the application.
+        /// </summary>
+        public ILogger Logger { get; }
+
+        /// <summary>
+        /// The list of addresses that the HTTP server is bound to.
+        /// </summary>
+        public ICollection<string>? Addresses => ServerFeatures.Get<IServerAddressesFeature>()?.Addresses;
```

Comment:
```suggestion
        public ICollection<string> Addresses => ServerFeatures.Get<IServerAddressesFeature>()?.Addresses ?? throw new InvalidOperationException("Server Addresses feature could not be found.");
```

Btw, we should super consider having a `GetRequiredFeature` and be done with this issue

Reply:
@Tratcher look your API 😄